validate file integrity

joosthooz opened the pull request:
resolves #1 
resolves #35 

Added a `validate` command that can:
- [x] => no longer doing this: _Verify the integrity of a dataset given by path_
- [x] Verify the integrity of all the entries in the cache

In addition; now checking checksums after downloading a file, if there is a checksum in the repo file

austin3dickey commented:
I'll review this once we merge #31 ; the diff is a little confusing otherwise

austin3dickey commented on .github/workflows/actions.yml:
```diff
@@ -119,8 +119,40 @@ jobs:
-      - name: Test generating and converting a dataset
+      - name: Test Downloading a dataset
         run: datalogistik generate -d type_floats -f parquet
+  test-download-convert-p2c-without-repo:
```
This new section feels a little redundant to me. I should have put more comments in this file, but what I was thinking for the CI strategy was:
- `lint`, `check-copyright`: self-explanatory
- `unit-tests`: test as much behavior of this package as possible, without calling the CLI
- `build-wheel -> test-convert-without-repo`: a simple integration test of the CLI without cloning the repo, in order to test that that workflow works
- `test-tpc-integration`: an integration test of TPC-H and TPC-DS because they're complicated and need additional setup than we could provide in `unit-tests`

If you want to test that the conversion works, I'd say your unit tests are good without this new section in this file. But if you wanted to be extra sure, you could also change line 123 to "csv" instead of "parquet". I don't think we need another whole step though.

joosthooz replied:
I understand, and removed the new test! Also changed output type to csv, to include a conversion. Maybe I'll add some more commands in this test, just so it hits most of the code

austin3dickey commented on datalogistik/cli.py:
```diff
@@ -31,6 +31,7 @@ def parse_args():
     sub_parsers = parser.add_subparsers(dest="command")
     cache_parser = sub_parsers.add_parser("cache")
     gen_parser = sub_parsers.add_parser("generate")
+    valid_parser = sub_parsers.add_parser("validate")
```
Could you please add documentation to the README? 🙂

joosthooz replied:
Yes! I Forgot (again), added

austin3dickey commented on datalogistik/util.py:
```diff
@@ -83,6 +84,73 @@ def add_file_listing(metadata, path):
+    metadata_file = pathlib.Path(path, config.metadata_filename)
+    dataset_found = False
+    if metadata_file.exists():
+        orig_file_listing = json.load(open(metadata_file)).get("files")
```
Could you please switch to the contextmanager `with open()` pattern here?

joosthooz replied:
Done, also for 1 spot in the unit tests

austin3dickey commented on tests/test_datalogistik.py:
```diff
@@ -101,8 +102,37 @@ def test_convert_dataset_csv_to_parquet():
     shutil.rmtree("./test_csv", ignore_errors=True)
+
+
+def test_convert_dataset_parquet_to_csv():
```
Thanks for this test! Could you add a test that the validation functionality works?

joosthooz replied:
Done in [88c8faf](https://github.com/conbench/datalogistik/pull/34/commits/88c8fafcaa550b85c462d0f61db1ac46dca3212f)

alistaire47 commented on datalogistik/util.py:
```diff
@@ -214,7 +291,7 @@ def get_dataset(input_file, dataset_info, table_name=None):
         po = csv.ParseOptions()
-        ro = csv.ReadOptions()
+        ro = csv.ReadOptions()  # autogenerate_column_names=True)
         co = csv.ConvertOptions()
```
If the source of truth is in csv which then gets converted on to parquet and whatever, these options can have a non-trivial impact on what gets written back out to other formats. [In benchmarks](https://github.com/voltrondata-labs/benchmarks/blob/main/benchmarks/_sources.py#L329-L354) I did a bit to lock it down, though I'd still hesitate to call it robust.

And for writing parquet [you may need to set `coerce_timestamps`](https://github.com/voltrondata-labs/benchmarks/blob/main/benchmarks/_sources.py#L432) for consistent timestamp precision depending on what parquet version you're using.

joosthooz replied:
Thanks, I'm actually struggling with these options because the schema detection in pyarrow.dataset is not good enough for the test datasets (fanniemae and nyctaxi). Passing the names would already help a bit but will not solve problems with the types of the schema: dataset detects the schema from the first x bytes of data (I think 1MB). So if there is a different value somewhere later on, it results in an error.
In the end a real solution would be to add the schema into the repo file and convert it to a pyarrow schema and pass it to the call creating the dataset. That's for #36 .

alistaire47 replied:
Yep, it's almost necessary to specify a schema when the original form is CSV/delimited; I ran into lots of issues with coerced nulls and null-type columns that caused downstream issues in addition to the obvious type differences

joosthooz replied:
I'm looking into coerce_timestamps, but `pyarrow.dataset` does not have en equivalent option.

alistaire47 commented on repo.json:
```diff
@@ -6,7 +6,7 @@
     "format" : "csv",
     "file-compression" : "gz",
-    "schema" : {
+    "tables": [{"table" : "fanniemae_2016Q4.csv", "schema" : {
```
[Here's a better schema for fanniemae](https://github.com/voltrondata-labs/benchmarks/blob/main/benchmarks/_sources.py#L54-L98) whenever you want it

joosthooz replied:
This is awesome, I'm going to add it right away

joosthooz replied:
[d3e9d58](https://github.com/conbench/datalogistik/pull/34/commits/d3e9d5827d019a733e6efb5b262e0012795182a7)

alistaire47 reviewed (approved):
Looks good! A few comments to highlight some landmines in data validation that I've discovered the hard way to be aware of as you move forward

joosthooz commented on datalogistik/util.py:
```diff
@@ -83,6 +84,73 @@ def add_file_listing(metadata, path):
 
 
+# Validate that the integrity of the files in the dataset at given path is ok, using the metadata file
+def validate(path):
```
1) Thanks for spotting this! I will take care of this.
2) Yes, also individual files will be placed in a directory, also because there will always be a metadata file added. The presence of that metadata file (and the "files" entry in it) will be the factor that determines whether the dataset seems valid (if the program finds a directory without it, it will delete the directory).
3) I don't check them because if the file size is different (so the files are not identical), you can bet your butt that the checksums are also different. But I can add a size check before the checksum check

joosthooz replied:
The deduplication is done

joosthooz replied:
Ok I looked at the integrity check code again, and it does check both file size and checksum, because it just checks the whole file listing array item (containing file path, size, and md5). This check is performed 1) after downloading and decompressing a dataset, but only in case there is a file listing for it in the repo file, and 2) at user request (either for all the entries in the cache, or for a given path)

alistaire47 replied:
Cool, size/checksum checks on the downloads should be pretty safe. Sizes get dicier for converted versions (e.g. a parquet/snappy/1e6 chunk size version of csv/gzip original) because they depend on all the reading and writing options, schema, and behavior, so any changes, deliberate or in defaults or behavior, cause failures. Less of an issue if deps are pinned, but I think the default parquet version is in the midst of changing, so choice of pin may be important

joosthooz replied:
Yes, checking generated/converted datasets will be very difficult. We would need to extract only the pure contents and have a checksum for that in the repo file too. That's probably a bit outside the scope of this tool.

jonkeane replied:
Yeah, agreed that the derived datasets is probably more complicated than it's worth (certainly for now possible forever). It also should be _relatively_ quick to re-generate the derived datasets anyway, such that if we're having issues we can prune the cache and then everything should be fine.

jonkeane commented:
Thanks for this! I haven't dug in too deeply, but I do wonder if we even need the ad-hoc verification command to be surfaced at all. In normal operations I would expect this to run after a download (which it is already), and that's probably about it. 

It's nice to have the ability to do this adhoc, but I wonder if it's adding more user interface than we really need to at this point?

joosthooz commented:
You may be right. I added it because some people might want to do read and writes to a dataset, but maybe that doesn't make sense and people should just re-instantiate in that case. If the dataset is in the cache that would be comparable to verifying. I'm ok with removing the command. Maybe we can just merge cache cleaning and verification into 1 option that does both?

joosthooz commented:
There is now no longer a validate subcommand, instead there is a cache --validate command that verifies file integrity of all the cache entries. Invalid entries are pruned.

jonkeane commented on README.rst:
```diff
@@ -9,7 +9,7 @@ generate some datasets such as TPC-H by calling their (external) generator progr
 Usage::
 
-    datalogistik <generate|cache>
+    datalogistik <generate|cache|validate>
```
```suggestion
    datalogistik <generate|cache>
```

If we only have an argument now we should do this, yeah?

austin3dickey replied:
Agreed; looks like you'll need to update the whole README again

jonkeane commented on datalogistik/cli.py:
```diff
@@ -44,6 +44,11 @@ def parse_args():
         action="store_true",
         help="Remove any incomplete/left-over directories from the cache",
     )
+    cache_group.add_argument(
+        "--validate",
+        action="store_true",
+        help="Validate all entries in the cache for file integrity",
+    )
```
If this auto-prunes, maybe we should call this `--prune-invalid` or something? That way it's clear what it's going to do if there's something that is invalid?

jonkeane reviewed (commented):
A few small comments and one big one: 

I thought the last time I looked at this, we had code that would validate the single source of data after it had been downloaded. But I'm struggling to find that again. Did we remove that? IMHO, validating after downloading is probably the only thing we want to expose right now (and do it unconditionally). Most datasets should be relatively easy to download such that if someone is concerned about their cache being corrupt, we should make it easy for them to delete everything for a datasource and start from scratch (downloading it).

Thoughts?

austin3dickey reviewed (changes requested):
README

joosthooz commented:
Sorry for forgetting the README again.
@jonkeane, validate is always performed if the repo file contains a file listing (with md5sums) for the downloaded dataset. But the location of that check was a bit strange (in `write_metadata()`), so I moved it to `download_dataset()` now.
I split the --validate option into --validate, that only reports if datasets are invalid, and --prune-invalid, that also deletes them.

austin3dickey commented:
>Sorry for forgetting the README again

haha no problem; once we do https://github.com/conbench/datalogistik/issues/25 we could have the docs automatically just print out the help page so we don't have to remember to update the README over and over

austin3dickey reviewed (approved):
I think this makes sense; I'm glad the validation is done upon download. I'm not positive anyone will ever need to use `--prune-invalid` or `--validate`, but if we don't want users to touch the cache directly then it makes sense to provide an approved way to do those things.
